src/mqtt_client.rs

```diff
@@ -365,27 +390,96 @@ where
                 Ok(())
             }
 
+            ReceivedPacket::PingResp => {
+                self.state.borrow_mut().pending_pingreq_time.take();
+                Ok(())
+            }
+
             _ => Err(Error::Unsupported),
         }
     }
 
+    fn open_socket(&mut self) -> Result<(), N::Error> {
+        let socket = self.network_stack.open(Mode::NonBlocking)?;
+        self.socket.borrow_mut().replace(socket);
+        Ok(())
+    }
+
     fn connect_socket(&mut self) -> Result<(), Error<N::Error>> {
+        // Don't try connecting more than once per second to avoid network spam if
+        // the broker port is closed.
+        if let Some(last_connect) = &self.last_socket_connect_time {
+            let now = self.clock.try_now().unwrap();
+            if now
+                .checked_duration_since(last_connect)
+                .and_then(|diff| Seconds::<u32>::try_from(diff).ok())
+                .map(|diff| diff < Seconds(1u32))
+                .unwrap_or(false)
+            {
+                return Err(Error::Disconnected);
+            }
+        }
+        self.last_socket_connect_time = Some(self.clock.try_now().unwrap());
```

Comment:
Unwrapping the clock is not safe. We need to propagate a clock error back out to the user. Ditto throughout the file.

Reply:
There's plenty of clocks that could be fallible (e.g. an external RTC over I2C)